dex/src/main/java/io/crate/data/BatchIteratorProxy.java

```diff
@@ -0,0 +1,131 @@
+/*
+ * Licensed to Crate under one or more contributor license agreements.
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.  Crate licenses this file
+ * to you under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied.  See the License for the specific language governing
+ * permissions and limitations under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.data;
+
+import io.crate.concurrent.CompletableFutures;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.function.Supplier;
+
+/**
+ * BatchIterator implementation that acts as a proxy for another BatchIterator which isn't immediately available.
+ *
+ * {@link #loadTrigger} will be used on the first {@link #loadNextBatch()} the request the BatchIterator to which
+ * it will delegate to once it's available.
+ */
+public class BatchIteratorProxy implements BatchIterator {
+
+    private final Supplier<? extends CompletionStage<? extends BatchIterator>> loadTrigger;
+    private final ColumnsProxy columnsProxy;
+
+    private BatchIterator delegate = null;
+    private CompletionStage<? extends BatchIterator> loadingFuture = null;
+
+    public static BatchIterator newInstance(Supplier<CompletableFuture<BatchIterator>> loadTrigger, int numColumns) {
+        return new CloseAssertingBatchIterator(new BatchIteratorProxy(loadTrigger, numColumns));
+    }
+
+    private BatchIteratorProxy(Supplier<? extends CompletionStage<BatchIterator>> loadTrigger,
+                               int numColumns) {
+        this.loadTrigger = loadTrigger;
+        this.columnsProxy = new ColumnsProxy(numColumns);
```

Comment:
i would rename this to LazyLoadingBatchIterator if that is the only intent, which i fear it is not,hm

the columns need to be there always, otherwise you cannot use the inputs and store them somewhere, before you start iterating